### ISSUE TITLE
JUnit XML logger does not handle `TestSuiteSkipped` event

### ISSUE THREAD

**Issue by thirsch** [sebastianbergmann/phpunit#5354]
| Q                   | A
| --------------------| ---------------
| PHPUnit version     | 10.1.2
| PHP version         | 8.2.5
| Installation Method | Composer

#### Summary

Mandatory JUnit-Report attribute `tests` in the first testsuite-Node is missing, if `setUpBeforeClass` uses e.g. `self::markTestSkipped()`. This is working fine with PHPUnit 9. As you can see in the examples, PHPUnit 9 handled the case also differently in the specific testsuite of the file containing the method call.

We are skipping the test class using `setUpBeforeClass` instead of using the `setUp` method, because there is a dataprovider generating many test cases that should not even be generated if skipped.

I've just changed our tests and just wanted to let you know what I've discovered.

#### Current behavior

```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites>
  <testsuite name="">
    <testsuite name="Application Test Suite" tests="0" assertions="0" errors="0" failures="0" skipped="0" time="0.000000">
      <testsuite name="SimpleTest" file="/src/tests/SimpleTest.php" tests="0" assertions="0" errors="0" failures="0" skipped="0" time="0.000000"/>
    </testsuite>
  </testsuite>
</testsuites>

```

#### How to reproduce

Use a simple test with the `self::markTestSkipped()` method call, e.g.:

```php
<?php 

class SimpleTest extends TestCase
{
    public static function setUpBeforeClass(): void
    {
        self::markTestSkipped('Skip all');
    }

    public function test(): void
    {
        self::assertTrue(true);
    }
}

```

#### Expected behavior

```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites>
  <testsuite name="" tests="1" assertions="0" errors="0" warnings="0" failures="0" skipped="1" time="0.000000">
    <testsuite name="Application Test Suite" tests="1" assertions="0" errors="0" warnings="0" failures="0" skipped="1" time="0.000000">
      <testsuite name="SimpleTest" file="/src/tests/SimpleTest.php" tests="1" assertions="0" errors="0" warnings="0" failures="0" skipped="1" time="0.000000">
        <testcase name="test" class="SimpleTest" classname="SimpleTest" file="/src/tests/SimpleTest.php" line="12" assertions="0" time="0.000000">
          <skipped/>
        </testcase>
      </testsuite>
    </testsuite>
  </testsuite>
</testsuites>
```

**Comment by asika32764** [comment]
Teamcity has same issue. This breaks phpstorm's test results.

![screenshot 2024-12-02 下午1 32 52](https://github.com/user-attachments/assets/5da319af-b960-493b-8e66-4d4d291588ed)

**Comment by asika32764** [comment]
I compare with phpunit 9

The correct output should contains `testIgnored` for test class and every test methods, and a `testSuiteFinished`.

```
##teamcity[testSuiteFinished name='Windwalker\Session\Test\Handler\MemcachedHandlerTest' flowId='57866']

##teamcity[testSuiteStarted name='Windwalker\Session\Test\Handler\XcacheHandlerTest' locationHint='php_qn:///private/var/www/windwalker/ww3/src/Session/Test/Handler/XcacheHandlerTest.php::\Windwalker\Session\Test\Handler\XcacheHandlerTest' flowId='57866']

##teamcity[testStarted name='testIsSupported' locationHint='php_qn:///private/var/www/windwalker/ww3/src/Session/Test/Handler/XcacheHandlerTest.php::\Windwalker\Session\Test\Handler\XcacheHandlerTest::testIsSupported' flowId='57866']

##teamcity[testIgnored name='testIsSupported' message='XCache is not enabled on this system.' details=' /private/var/www/windwalker/ww3/src/Session/Test/Handler/XcacheHandlerTest.php:36|n ' duration='0' flowId='57866']

##teamcity[testFinished name='testIsSupported' duration='0' flowId='57866']

...

##teamcity[testSuiteFinished name='Windwalker\Session\Test\Handler\XcacheHandlerTest' flowId='57866']

```

At phpunit >= 10, it only contains `testSuiteStarted`, no `Finished` log, so phpstorm output will broke.

![screenshot 2024-12-02 下午3 55 25](https://github.com/user-attachments/assets/0136b93c-caa0-4128-9a2e-e39572578635)

**Comment by sebastianbergmann** [comment]
Any comparison with PHPUnit 9 is pointless as PHPUnit >= 10 works completely different internally.

**Comment by asika32764** [comment]
Yes but I'm not comparing about the code, I think the difference about the logging output is the main cause for this issue.

**Comment by sebastianbergmann** [comment]
When `markTestSkipped()`, for example, is used in a before-class method such as `setUpBeforeClass()` then the individual test methods of the class in question are not processed and no events are emitted for them. This is why the expected behavior you describe is not possible.

This woud be possible:

```xml
<testsuite name="..." file="...">
    <skipped/>
</testsuite>
```

However, I do not know if the above would be valid JUnit XML.

As I have said over the years over and over again: I am very reluctant to make changes to the JUnit XML logger as the is no real standard for it, no canonical schema we can validate against, and each CI tool interprets it differently.

**Comment by sebastianbergmann** [comment]
When `markTestSkipped()` is used in a before-class method such as `setUpBeforeClass()`, the `TestSuiteSkipped` event is emitted (instead of individual `TestSkipped` events for each test method). The JUnit XML logger did not subscribe to this event. As a result, the `<testsuite>` element for the skipped class had no `tests`, `skipped`, or `time` attributes and its test count was not propagated to parent suites.

I have implemented a fix for PHPUnit 12.5 and PHPUnit 13.1: When a test suite is skipped, the logger now increments the `tests` and `skipped` counters by the number of test methods in the suite and then finalizes the `<testsuite>` element (since `TestSuiteFinished` is not emitted for skipped suites).

As I have said many times over the years: I am very reluctant to make changes to the JUnit XML logger. There is no canonical specification for this format. There is no schema we can validate against. And each CI tool interprets the format differently. This is why the fix is intentionally minimal: it only updates counters and does not introduce synthetic `<testcase>` elements for tests that were never actually executed.

If you rely on JUnit XML for reporting, please consider migrating to [Open Test Reporting XML](https://github.com/ota4j-team/open-test-reporting). PHPUnit has supported this format since [PHPUnit 12.2](https://github.com/sebastianbergmann/phpunit/releases/tag/12.2.0). It is a well-specified, modern alternative with a proper schema. Please encourage your CI/CD tool vendors to implement support for Open Test Reporting XML in addition to JUnit XML. At some point in the future, support for JUnit XML logging in PHPUnit will be discontinued.
